### ISSUE TITLE
Is linter currently really useful?

### ISSUE THREAD

**Issue by vvs-** [LostArtefacts/TRX#1062]
>>> While I agree that this linter has many shortcomings, it still brings enough value to the project to keep it. As far as I know, there are currently no superior alternatives available for C. At least the suggested modification it proposes is relatively minor.

>> I'm sorry, With all respect, I'd argue that asking random contributors to fix long standing issues using their own unrelated changes just as an excuse is not a good idea. Running linter separately from all other activity, for example, might be better.

>> That said, I've managed to fix that warning which I have no idea why it even happened in the first place. But other people might not be so lucky.

> If the linters are isolated and not integrated into our regular development workflow, they may be overlooked or forgotten, defeating their purpose of catching issues early on.

> I perceive your behavior to be somewhat rude – I am not requesting anything from you; it is my choice to dedicate my personal time away from my family to address this pull request and enable you to compile in a way that is not officially supported. Additionally, you indicated that you have read the contributor guidelines, but you have disregarded our recommended commit message format. From now on please be more mindful of the way you frame your messages. Thanks.

This was not a good way to start a honest discussion. But I'll try to change this here.

First of all, I'm _politely_ wondering if the current linter behavior is really that useful for beginners? I'm not telling anyone else what they _should_ actually do, that was just a (naive) suggestion. Right now any drive-by contributions might be hindered by _unrelated_ failures from linter which might be somewhat cryptic to unsuspecting person.

Second, I did not ask anyone to let me compile anything in unsupported way. I saw a broken code which I thought would be beneficial for the project to fix for everyone. Personally, I've already fixed it locally so I didn't need any further "enablement" to do so on my own. I saw this as an opportunity to do a favor for a community, not for myself. Just because an official build already has the root directory in its path for includes doesn't necessarily mean that this was intentional.

Also, If I did break guidelines for making a contribution, I'm sorry. This was not intentional. I'm working with a different code bases all the time and just didn't have enough time to check that every small nit was in order, especially in a very trivial one-liner. BTW anyone could point me to my mistake without assuming malicious intent on my part. In fact, last time I contributed code here, someone just squeezed my commits, replacing their subject without complains. Outright blaming without even giving me a chance to defend myself is not nice either.

Now back to the subject. I did see that previously other contributors had difficulties with submitting their changes due to linter complaining in a non-obvious way. I can imagine people trying to fix a small typos and facing requests from linter to fix some other parts of code which they didn't ever touched or understand. I thought that at least pointing that out is an admittance of their grievances. If nobody is going to do anything about it, that's fine with me. I didn't try to enforce my POV on anyone else here, it's a honest question.

**Comment by rr-** [comment]
>Right now any drive-by contributions might be hindered by unrelated failures from linter which might be somewhat cryptic to unsuspecting person.

I understand your point and I agree. In the case of IWYU, eg the linter in question – its main purpose is for us to avoid `#include` bloat and only include what we use. It does sometimes feel buggy and its suggestions to be unnecessary, but they never produce wrong code, and most of the time, its suggestions are sensible and do improve the general quality.

Potential ways to address this:

1. Replace IWYU with something better. However, as far as I know there are no alternatives on the market for the C ecosystem.
2. Remove IWYU and possibly all linters altogether. That is not going to happen as these are a big time saver for us.
3. Postpone linting until after the PR is merged. This has a number of drawbacks:
    - we'll be merging nonconforming code to `develop` and this feels dirty.
    - it's not clear who should clean it up after merging.
    - the feedback comes late, forcing us to do lots of "fix style" commits.
4. Change the CI job not to flag nonconforming commits, and instead to automatically send commits that fix the perceived wrongdoings. Again, this leads to history pollution, but that can be mitigated by squashing merges. More importantly, I think this may reduce contributors engagement, as whenever they are requested to do changes and try to push, they'll discover that upstream has diverged and has additional automated commits – in their own branch, in their own fork. This dulls sense of ownership of their own contributions.
5. Have the reviewer push the style fix manually to the PR. I sometimes do this, but it suffers from the same drawbacks as above.
6. Improve the feedback from the linters, which now only output a patch what changes they deem to be necessary. This may prove difficult.

I'm leaning towards point 4. but I'm not sure how other regular contributors would feel about this.

**Comment by lahm86** [comment]
I agree the benefits of IWYU outweigh the negatives.

I used to see differences between running IWYU locally and the output generated at PR time. But since I updated my system some months ago, these no longer occur. I am on Ubuntu 23.04, clang-15, and IWYU 8.18-2 (already out of date, yes, but previously I was on much older versions).

I don't really like the idea of automatic commits. I think it's good the way it currently is, as in where the CI flags issues and the contributor reviews.

What we could maybe do to improve things is:

- Show in the IWYU output the version that is being used, so if contributors are seeing differences on their local system, that could be the first thing to check
- Make a combined `make lint` and `make imports` tool so contributors need only run one command before committing
- Update `CONTRIBUTING.MD` to mention above - we talk about lint but not really imports explicitly, so that could be of benefit
